src/main/scala/ch/epfl/bluebrain/nexus/kg/async/Cache.scala

```diff
@@ -0,0 +1,58 @@
+package ch.epfl.bluebrain.nexus.kg.async
+
+import java.util.UUID
+
+import akka.actor.ActorSystem
+import cats.Monad
+import cats.effect.{Async, Timer}
+import cats.implicits._
+import ch.epfl.bluebrain.nexus.kg.resources.Rejection
+import ch.epfl.bluebrain.nexus.kg.resources.Rejection.Unexpected
+import ch.epfl.bluebrain.nexus.service.indexer.cache.KeyValueStore.Subscription
+import ch.epfl.bluebrain.nexus.service.indexer.cache.{KeyValueStore, KeyValueStoreConfig, OnKeyValueStoreChange}
+
+abstract class Cache[F[_]: Monad, V](private[async] val store: KeyValueStore[F, UUID, V]) {
+
+  /**
+    * Adds a subscription to the cache
+    *
+    * @param value the method that gets triggered when a change to key value store occurs
+    */
+  def subscribe(value: OnKeyValueStoreChange[UUID, V]): F[KeyValueStore.Subscription] = store.subscribe(value)
+
+  /**
+    * Removes a subscription from the cache
+    *
+    * @param subscription the subscription to be removed
+    */
+  def unsubscribe(subscription: Subscription): F[Unit] = store.unsubscribe(subscription)
+
+  private[async] def get(id: UUID): F[Option[V]] = store.get(id)
+
+  private[async] def replace(id: UUID, value: V): F[Unit] = store.put(id, value)
+}
+
+object Cache {
+
+  private[async] def storeWrappedError[F[_]: Timer, V](
+      name: String,
+      f: V => Long)(implicit as: ActorSystem, config: KeyValueStoreConfig, F: Async[F]): KeyValueStore[F, UUID, V] =
+    new KeyValueStore[F, UUID, V] {
```

Comment:
it seems that the implementation of the `KeyValueStore` is not flexible enough to use as is, we should probably add an error mapper to avoid this subclassing
(not now though)